### PR TITLE
Detailed asset lock test

### LINKED ISSUE #14: Unit tests for TimedAssetLockContract that  are based off local state/order of operations

### PR DESCRIPTION
**Pull request by palmerss**

### CODE REVIEW

File: test/timed_asset_lock_contract_test.py [hunk 1]
```diff
@@ -79,112 +89,300 @@ def app_id(test_config, asset_id, end_time, wallet_1):
...
+                                                 params,
+                                                 app_id,
+                                                 [asset_id])
+    if adversary_wallet or fail_clear:
+        with pytest.raises(AlgodHTTPError):
+            client.send_transactions([txn])
+            wait_for_txn_confirmation(client, txn_id, 5)
+
+
+def cash_out(client, public_key, private_key, app_id, asset_ids):
```
Comment (drewdrew12345) [inline test/timed_asset_lock_contract_test.py:189]: should these functions be in testing_utils.py?

cash_out, opt_out, opt_in, set_up

File: test/timed_asset_lock_contract_test.py [hunk 2]
```diff
@@ -1,60 +1,69 @@
...
+from akita_inu_asa_utils import read_local_state, read_global_state, wait_for_txn_confirmation
+
+
+NUM_TEST_ASSET = int(1e6)
+ESCROW_TIME_LENGTH = int(90)
+
 
 @pytest.fixture(scope='class')
 def test_config():
     from test.testing_utils import load_test_config
```
Comment (drewdrew12345) [inline test/timed_asset_lock_contract_test.py]: In the other cases the import is from testing_utils not test.testing_utils - is there a reason for having both?

File: contracts/timed_asset_lock_contract/program.py [hunk 3]
```diff
@@ -150,28 +154,30 @@ def closeAccountTo(account: Expr) -> Expr:
...
 
 
 def clear_program():
-    return (compileTeal(Approve(), Mode.Application, version=5))
+    return compileTeal(Reject(), Mode.Application, version=5)
 
 
-def compile(algodClient):
+def compile_app(algod_client):
     dump_teal_assembly('assetTimedVault_Approval.teal', approval_program)
```
Comment (drewdrew12345) [inline contracts/timed_asset_lock_contract/program.py]: asset_timed_vault_approval.teal unless this is standard for teal files

File: akita_inu_asa_utils.py [hunk 4]
```diff
@@ -47,8 +52,26 @@ def compile_program(client, source_code, file_path=None):
         dump(base64.b64decode(compile_response['result']), 'build/' + file_path)
 
 
+# read user local state
+def read_local_state(client, addr, app_id):
+    results = client.account_info(addr)
+    print(results)
```
Comment (drewdrew12345) [inline akita_inu_asa_utils.py]: can prob remove this print statement

File: akita_inu_asa_utils.py [hunk 5]
```diff
@@ -47,8 +52,26 @@ def compile_program(client, source_code, file_path=None):
...
 
 
+# read user local state
+def read_local_state(client, addr, app_id):
+    results = client.account_info(addr)
+    print(results)
+    if len(results['apps-local-state']):
+        local_state = results['apps-local-state'][0]
+        for index in local_state:
+            if local_state[index] == app_id:
```
Comment (drewdrew12345) [inline akita_inu_asa_utils.py:63]: if local_state[index] == app_id and if 'key-value' in local_state.keys():

File: akita_inu_asa_utils.py [hunk 6]
```diff
@@ -173,6 +196,25 @@ def create_app_signed_txn(private_key,
     return signed_txn, signed_txn.transaction.get_txid()
 
 
+def update_app_signed_txn(private_key,
```
Comment (drewdrew12345) [inline akita_inu_asa_utils.py:201]: indicate in comment that function returns tuple of signed_txn and txid

same for other app functions here

File: test/testConfigExample.json [hunk 7]
```diff
@@ -3,6 +3,5 @@
   "algodToken": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
   "indexerAddress": "http://localhost:8980",
   "indexerToken": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
-  "fund_account_public_key": "DO NOT STORE YOUR REAL KEY HERE",
-  "fund_account_Mnemonic": "DO NOT STORE YOUR REAL KEY HERE"
+  "fund_account_Mnemonic": "DO NOT STORE YOUR REAL KEY HERE",
```
Comment (drewdrew12345) [inline test/testConfigExample.json]: fund_account_mnemonic and update references to it - not really important though

File: test/timed_asset_lock_contract_test.py [hunk 8]
```diff
@@ -1,60 +1,69 @@
...
     from akita_inu_asa_utils import generate_new_account
-    wallet_mnemonic, private_key, public_key = generate_new_account()
     from testing_utils import fund_account
-    wallet_1 = {}
-    wallet_1['mnemonic'] = wallet_mnemonic
-    wallet_1['publicKey'] = public_key
-    wallet_1['privateKey'] = private_key
+    wallet_mnemonic, private_key, public_key = generate_new_account()
+
+    wallet_1 = {'mnemonic': wallet_mnemonic, 'publicKey': public_key, 'privateKey': private_key}
```
Comment (drewdrew12345) [inline test/timed_asset_lock_contract_test.py]: public_key, private_key and update references to these

File: test/timed_asset_lock_contract_test.py (hunk 1 above)
Comment (palmerss) [reply test/timed_asset_lock_contract_test.py:189]: I don't believe so, as these are targeted for use of this smart contract specifically

File: akita_inu_asa_utils.py (hunk 4 above)
Comment (palmerss) [reply akita_inu_asa_utils.py]: oops leftovers from debugging

File: akita_inu_asa_utils.py (hunk 5 above)
Comment (palmerss) [reply akita_inu_asa_utils.py:63]: yup! Depending on the condition of the local state. If I remember right, local_state[index] may contain the local schema, however does not guarantee that the 'key-value' key exists, which contains local_state variables

**Review by drewdrew12345 — APPROVED** [review]
Lgtm

**Pull request merged**